### PR TITLE
Filter rules gui

### PR DESCRIPTION
**Pull request by CautemocSg**
Adds an in-extension user interface for editing and adding filter rules previously handled by the JSON editor

### CODE REVIEW

**Comment by errnoh** [comment]
Doesn't seem to open in Firefox (probably good idea to check other browsers as well, for example there are bunch of users using some android browsers)

File: src/components/RuleBasedFilters/RuleBasedFilters.tsx [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+              <div>
+                <StyledLabel htmlFor="deemphasize-by">
+                  <RuleText size={Size.Medium}>De-emphasize: </RuleText>
+                </StyledLabel>
+                <select
+                  id="deemphasize-by"
+                  value={props.DE}
+                  onChange={(event) => {
+                    props.setDE(event.target.value as DeemphasizeOption)
+                    localStorage.setItem("deemphasize-selection", event.target.value)
```
Comment (danreeves) [inline src/components/RuleBasedFilters/RuleBasedFilters.tsx]: Local storage access was moved to a `useLocalStorage` hook which namespaces our keys

File: src/types.ts [hunk 2]
```diff
@@ -223,5 +226,21 @@ export interface FilterRule {
...
+  minStats: string
+  minRating: string
+  isOpen: boolean
+}
+
+export enum Size {
+  Small,
+  Medium,
+  Large
+}
```
Comment (danreeves) [inline src/types.ts]: Not a fan of enums, this could be a string union

**Review by danreeves — COMMENTED** [review]
I'm not a fan of adding such a giant client side dependency as styled-components. What do you think about using something like linaria where we can generate the css at build time https://github.com/callstack/linaria

File: src/components/RuleBasedFilters/RuleBasedFilters.tsx (hunk 1 above)
Comment (CautemocSg) [reply src/components/RuleBasedFilters/RuleBasedFilters.tsx]: Fixed, re-implemented as useLocalStorage

File: src/types.ts (hunk 2 above)
Comment (CautemocSg) [reply src/types.ts]: Fixed, re-implemented as string union

**Comment by CautemocSg** [comment]
> Doesn't seem to open in Firefox (probably good idea to check other browsers as well, for example there are bunch of users using some android browsers)

It was broken in Firefox, latest should be working now. Tested on my desktop is working as intended. Could use a check on phone browsers.

**Comment by errnoh** [comment]
Yup, seems to be working on Firefox now.

I initially had some problems understanding the save button is at top. If you have bunch of rules you'll need to scroll upwards to find the save button. Should it be at the bottom?

positive:
+ I like that it updates JSON instantly when saving
+ UI themed with the site :)

negative:
- JSON is a bit verbose with all fields being visible with default zero values 
- Not yet on feature parity with the current available rbf rules
- The rule string doesn't seem to show everything, I have some rules that show as empty even when they have some rules defined. (e.g. rule with only blessing rarity defined)

**Comment by errnoh** [comment]
And a feature wish: allow dragging rules to reorder them. Wasn't relevant previously, but with the `color` option it's a proper QoL improvement :)

**Comment by CautemocSg** [comment]
Addressed these points with recent update:

> * JSON is a bit verbose with all fields being visible with default zero values
> * Not yet on feature parity with the current available rbf rules
> * The rule string doesn't seem to show everything, I have some rules that show as empty even when they have some rules defined. (e.g. rule with only blessing rarity defined)

**Comment by errnoh** [comment]
Great changes! Hopefully I wasn't too blunt with my feedback :) Just wanting to be sure that such a major UI overhaul feels intuitive to the users.

**Comment by errnoh** [comment]
Couple additional things that might be good features. Maybe not worth scoping in on this PR, but related:
- Have the highlight color visible in the rules listing somehow (not just edit window). Small box with that color? Text on the highlight color? I've been locally using the coloring rules for some time already and while it has been one of the largest QoL improvements I still sometimes forget what each color actually means, I'd assume others will run into the same issue.
- Autocomplete / validation. For some fields it's quite simple as the possible values are limited by the type, for some we could do suggestions based on data available (blessing/perk names etc). Then again for weapons there are a lot of cases where one might not want to use full item names, while for blessings that's actually usually the case.

**Comment by CautemocSg** [comment]
>And a feature wish: allow dragging rules to reorder them. Wasn't relevant previously, but with the color option it's a proper QoL improvement :)

>Have the highlight color visible in the rules listing somehow (not just edit window). Small box with that color?

@errnoh  Thanks for the feedback and suggestions.

I implemented these 2 requests in the latest update.

**Comment by errnoh** [comment]
Replied on Discord but I'll add it to discussion here as well:

👍 from me functionality-wise. It's in a state that I have no need to manually edit JSON other than to test things and share configs between browsers & machines. I think if we can get someone to thumbs up the code changes this should be ok to merge?

Btw I've tested this on Firefox and Chrome. If someone has a setup where they can test that it works fine with the android browser bunch of people are using that would be a good thing to check.

**Review by danreeves — APPROVED** [review]
This is awesome!

**Pull request merged**